src/Plugin/RulesExpression/RulesLoop.php

```diff
@@ -91,4 +92,46 @@ public function checkIntegrity(ExecutionMetadataStateInterface $metadata_state)
     return $violation_list;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function prepareExecutionMetadataState(ExecutionMetadataStateInterface $metadata_state, ExpressionInterface $until = NULL) {
+    if ($until && $this->getUuid() === $until->getUuid()) {
+      return TRUE;
+    }
+
+    $list_item_name = isset($this->configuration['list_item']) ? $this->configuration['list_item'] : 'list_item';
+    try {
+      $list_definition = $metadata_state->fetchDefinitionByPropertyPath($this->configuration['list']);
+      $list_item_definition = $list_definition->getItemDefinition();
+      $metadata_state->setDataDefinition($list_item_name, $list_item_definition);
```

Comment:
The list item is removed in both cases, so this works as designed. Added a test case to also make sure of that.